source/hpack/encoder.d

```diff
@@ -0,0 +1,265 @@
+module hpack.encoder;
+
+import hpack.tables;
+import hpack.huffman;
+import hpack.util;
+import hpack.exception;
+
+import vibe.http.status;
+import vibe.http.common;
+
+import std.range;
+import std.bitmanip; // prefix encoding / decoding
+import std.typecons;
+import std.conv;
+import std.exception;
+import std.array;
+
+struct HeaderEncoder(T = HTTP2HeaderTableField[])
+	if(isInputRange!T && is(ElementType!T : HTTP2HeaderTableField) ||
+	   is(T == HTTP2HeaderTableField))
+{
+	private {
+		static if(is(T == HTTP2HeaderTableField)) T[] m_range;
+		else T m_range;
+		IndexingTable m_table;
+		ubyte[] m_encoded;
```

Comment:
It looks like supporting the input range interface, while still aiming for `@nogc`, is rather difficult due to the complex nature of the stream (e.g. dynamic length header/value + dynamic length Huffman encoding). With a few assumptions and a more complex encoder state it would certainly be possible, but for practical reasons it may make sense to concentrate on the output range based API.